proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1417 @@
+
+     hr :: (forall a. a -> a -> a) -> ...
+     hr = ...
+     g = hr (\ x y -> ...)   -- there must be some way to name the type of x or y here
+
+   Once we have the EVP, there will never be a need for ``Proxy``.
+
+   Motivation: As GHC supports more and more type-level programming, the ability
+   to write out type signatures, arguments, and annotations has become increasingly
+   important. With ``-XScopedTypeVariables``, GHC allows us to bring type variables
+   into scope, but often requires us to do so by cumbersome matching. If we have
+   a type ``Maybe (Either (Int, Bool, a) Double)``, that's a lot to type just to
+   be able to, say, bind ``a``. The EVP says we do *not* have to resort to matching,
+   ever.
+
+   .. _EBP:
+
+#. **Explicit Binding Principle (EBP)**. Through the right combination of extensions and/or warning flags, it is possible
+   for a Haskell programmer to ensure that all identifiers in a program have an explicit binding site.
+
+   Examples::
+
+     id :: a -> a    -- the variable `a` has no explicit binding site, but we can write `forall a.` to provide one
+
+     f :: (Bool, Bool) -> Bool
+     not (x :: (b, b)) = ...   -- the variable `b` is bound to `Bool` by this
+                               -- pattern signature. But either this is done by
+                               -- matching (in violation of the EVP) or the binding
+                               -- site is implicit (in violation of the EBP).
+
+   Motivation: The EBP allows programmers to control exactly how variables come into
+   scope. It also prevents the possibility of typos that accidentally introduce new
+   variables.
+
+   .. _VOP:
+
+#. **Visibility Orthogonality Principle (VOP)**. Whether an argument is visible or
+   invisible should affect only its visibility, not other properties.
+
+   A consequence of the VOP is that these two programs should have the same meaning::
+
+     f1 :: forall a -> ...
+     f1 blah1 = ...
+
+     g1 = ... f1 blah2 ...
+
+     -------
+
+     f2 :: forall a. ...
+     f2 @(blah1) = ...
+
+     g2 = ... f2 @(blah2) ...
+
+   The only difference between these is the visibility.
+
+   Motivation: Visibility should be just that: a superficial property that describes
+   (only) whether an argument is visible in the user-written source code.
+
+   .. _PEDP:
+
+#. **Pattern/Expression Duality Principle (PEDP)**. The syntax for patterns mimics
+   that of expressions, allowing an expression headed by a constructor to be pattern-matched
+   against a pattern of the same syntactic structure.
+
+   Motivation: This is the essence of pattern-matching, where we can deconstruct data
+   that was constructed by an expression.
+
+.. _universals-and-existentials:
+
+Universals and existentials
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The combination of the VOP_ and the LSPC_ lead to a surprising conclusion: we must distinguish
+between universal variables and existential variables of GADT constructors. This section explains
```

Comment:
Hmm, I am not sure I fully buy into this argumentation.

The `VOP` and the `LSPC` merely imply that binders and occurrences of variables need _syntactic_ distinguish, independent of the type of the matched constructor. But, precisely because of the LSPC, there can’t a syntactic distinction between the `p1 and the `p2` in `MkUE p1 p2`. So any distinction we need, if any, may be a non-syntactic.

Yes, we seem to want to revert the solution in the types-in-patterns paper of using the in-scope set to distinguish, and we (eventually) need syntax. But I don’t see how that invalidates the other conclusions; in particular, that there is _no_ distinguish between existentials and universals. For example, I am still content to have, say,

```
data FakeEx a where MkFakeEx :: forall a b -> a ~ b => FakeEx a
foo :: FakeEx Int ->  …
foo (MkFakeEx _ Int) = …
```
or even fun stuff like the following, where we use non-trivial pattern in what is _certainly_ an existential to match on the existential type (which GHC knows something about) and then bind something inside:
```
data ExWithC (c :: * -> Constraint) where MkExWithC :: forall a -> c b => ExWithC c
type family UnMaybe a where UnMaybe (Maybe a) = a
type IsMaybe a = a ~ Maybe (UnMaybe a)
foo :: ExWithC IsMaybe ->  …
foo (MkExWithC _ (Maybe a)) = …
```